### ISSUE TITLE
Setting for moving body of function to next line?

### ISSUE THREAD

**Issue by knocte** [fsprojects/fantomas#848]
I was a happy user of `--keepNewlineAfter`, and now that it's been removed, I feel there should be a way, at least, to tell fantomas to always start the body of a function (or a member, in case the function is a method of a class) always in a new line. In my view, it improves readability to only let constants (and not functions) to be declared and assigned in the same line. This way by a quick glance you can most of the time know if you're looking at a constant or at a function.

Could a setting be created for this? (I'm not sure about the best name for it though, `functionBodyToStartAfterNewLine`?) If yes, I will contribute it.

**Comment by nojaf** [comment]
Hey @knocte, I understand your sentiment here and I feel a setting could be ok here.
However, home work first ;)
Could you identify a code example for all different cases where the setting would have impact.
Please come up with an excessive list of examples. Think about subtle differences like signature files versus implementation files. Top level let bindings versus let bindings within let bindings.

**Comment by nojaf** [comment]
Once we have identified the cases, maybe the preferred behavior can also be achieved by introducing some settings that control the max length of expression.

Similar how you can now push down even the smallest value by setting [MaxLetBindingWidth to 0](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMBDMAvGACyQA04UAztlgJYB2A5lamANqUSgS9QAkWAJ4AHZGkiMs5Lr0j8AbrnQBXcR1l8ADBT59IASQZIGWAMojc0RADkVAWxl7exTWAC6sgL67ePPfLCYmySpk4BispqbJzOEACMvs6QAAq4TIgA6nTwWAAW4c7xAExabp58PrL%2B%2BoKi6pAARrCw6IVykarqsXHFSQFmiPZ0cOiwDACCWACixgDyAGYAMozI-XwLyjTl3v01HUENza3tEPJKXTFuEADM63IWVogAQogLsABOiGkfuPbYiA%2Bp14WA%2Bah2lT2bkC9RCxza93OUW61xI9zOj2sr3eXyWsAA7oDoLgaEYFLBifRxsCIKDwc4Krwqnx9mc6sEJPDgUjLmges4AKzoqCYl5vT6IACqIjEH2JpIY5MpdGpws26G2DN21Wh7KOLQRuou0T5qIAbMLIKLsRKAMLoElUW3jKh06BYT40sDqzV6RkQZl%2BXWHOEG7mdE0aOJgADslut4q%2BAFkho1AV6fYgIUyock9aGToiIyjowAOeOWLGJxDOsYMDNbLNayE6vMhzlhovGktxACcFaeEwWWEBzvs9lwXrpTb92pZwdhHcLRuRV2j8R0qKtlbFOMQg2Go1VqMz2YDuYi7agXK7q9N68SW9FQ5HHwPI1ax%2Bj07PYED3AXDlr07FdeSjOISktIwTCwOYGAAFQ%2BIRsnyBsNRnPh-T-C9aivJoQLzbs1wgu4nx3CYPlgFRjAAEUQdA6GGV8pzBDDeCw-8wFZAQ8KkcMiPvCDiEtJNcAADwMBZ4LyRAGGmdCzDyT4sGyXICmFYgymbHNW0vRcoD428wP5PR4iFLdRIkhgFjoMS5llXAPQ%2BaYxJEL4qCoFV62FAUtNnFt5zbfTQmkIzIxMvh4gtCzxIAJUQOAPngVTUI0vzMLnIMgqAkL%2BLvcCijjGKxIo34hDmD4VldFL1NRTTf047iYRywzQPC1F4nLYqlmwZ5GHgRgmBqr10vYzKAOy-Vl0I-KIt4eJ%2BwslR0HoBiGBeMZoAAa2eX5tuwKg4LMP4a1aBxvJPRsGpwg5gpvNqe2cUpLRsRB8TWsUsEJWT4PqOibIYOgqUmYwU3sNMPlYNUru089dNwu6CIiASCr0YpH2jSAAGlEEQERJOk2SjGO-4VnWtDfQygKsr0nL7pm4zUWKPon1BEYsCTBA1ku9CGsodxfEkKgADFPLYU8vCAA).

**Comment by knocte** [comment]
The basic samples are:

With setting disabled:
```
let ShowSaneDate (date: DateTime): string = date.ToString "dd-MMM-yyyy"
```
With setting enabled:
```
let ShowSaneDate (date: DateTime): string =
    date.ToString "dd-MMM-yyyy"
```

With setting disabled:
```
member internal self.GetEncryptedPrivateKey () = accountFile.Content ()
```
With setting enabled:
```
member internal self.GetEncryptedPrivateKey () =
    accountFile.Content ()
```

I'll let @Bobface write the additional convoluted samples :)

**Comment by knocte** [comment]
> Similar how you can now push down even the smallest value by setting MaxLetBindingWidth to 0.

MaxShortFunctionBindingWidth?

**Comment by nojaf** [comment]
Maybe `MaxMemberBindingWidth` as so far we only are missing a setting for the member bindings?

**Comment by knocte** [comment]
> as so far we only are missing a setting for the member bindings

Are you trying to say that if I hadn't the same desire to apply this setting to members, I would have enough with MaxLetBindingWidth? If yes, then you didn't understand that I was only referring to functions, not constants (so if I use MaxLetBindingWidth=0, even things like `let x = 0` get divided in 2 lines, which was not my aim).

**Comment by nojaf** [comment]
Ah, I see. Yes I was approaching this more from an AST perspective.
So in the case of
```fsharp
let a = 7
let add b = a + b
```
with MaxLetBindingWidth = 0 we now go to:
```fsharp
let a =
    7

let add b =
    a + b
```

if I understand correctly you'd like to have that as:
```fsharp
let a = 7

let add b =
    a + b
```

It could make sense to split MaxLetBindingWidth into two settings.
Maybe LetBinding is a bit too broad.

**Comment by knocte** [comment]
> if I understand correctly you'd like to have that as:

Exactly.

> It could make sense to split MaxLetBindingWidth into two settings.
> Maybe LetBinding is a bit too broad.

But if we split MaxLetBindingWidth we would have only taken care of half of the problem here (members left). You want to have 4 settings? MaxValueLetBindingWidth, MaxFunctionLetBindingWidth, MaxValueMemberBindingWidth, MaxFunctionMemberBindingWidth? Seems a bit excessive to me.

**Comment by knocte** [comment]
I'd still prefer a one single extra setting, such as `MaxShortFunctionBindingWidth(int)` or `FunctionBodyToStartAfterNewLine(bool)`. @nojaf let me know your choice

**Comment by nojaf** [comment]
The main problem I have with `FunctionBodyToStartAfterNewLine` is that it has overlap with `MaxLetBindingWidth`.
What should happen if a user tries to format
```fsharp
let add a b = a  + b
```
with settings:
```json
{ 
     "MaxLetBindingWidth": 1000,
     "FunctionBodyToStartAfterNewLine": true
}
```
What setting should it respect? Even if you choose one of the two settings as the victor, you need to take two settings into account to get a result. That will lead to confusing and something I also rather not have in the codebase. People will open bugs to subverted expectations.

That is why I'm more in favour of small settings that don't cross each other boundaries.

**Comment by knocte** [comment]
> What setting should it respect?

I don't think setting FunctionBodyToStartAfterNewLine to true is contradicting a high number for MaxLetBindingWidth, it would just set an exception for which let bindings you want to apply the width to.

Anyway, let's try to move this forward. If you prefer the setting split idea, my main concern about it is generating so many settings out of it (because members, as opposed to lets, would need settings too). How about we do it in 2 phases? Deprecate MaxLetBindingWidth in favour of a more generic one: MaxBindingWidth (that covers lets and members); after that, separate it into two: MaxFunctionBindingWidth and MaxValueBindingWidth.

**Comment by nojaf** [comment]
In terms of actual work, I don't think it is that much more work to have the split into `MaxFunctionBindingWidth` and `MaxValueBindingWidth` right away.

But I understand your sentiment and we could go for a phased approach where you do the work to rename MaxLetBindingWidth to MaxBindingWidth and have it also use it for member bindings.
Someone else could pick up the next phase of splitting MaxBindingWidth into MaxFunctionBindingWidth and MaxValueBindingWidth.

Another side note, just so we are on the same page, when you mention MaxValueBindingWidth.
The different between Function and Value is determined purely on the first part between `let` and `=` right? That is how I'm interpreting this and so things like
```fsharp
let a = someA + someB + someC
```
are also consider to follow the rule of MaxValueBindingWidth right?

**Comment by knocte** [comment]
> The different between Function and Value is determined purely on the first part between let and = right? 

Yes. If there are more than 1 element between the let and the `=` sign, then it's not a value but a function (so it doesn't get evaluated right away, only when it's called).

**Comment by knocte** [comment]
AFAIU this has been fixed with 50ed1b7aab9bde89ea4d5d3fc6fd44b91f0741cb.
